tests/unittests/AddressBook/AuthorizationTests.mm

```diff
@@ -0,0 +1,143 @@
+ * mocked version of WDEDeviceAccessInformation, and returns the original
+ * implementation of createFromDeviceClassId:.
+ */
++ (IMP)setupMock;
+
+/**
+ * Swizzles the given implementation to WDEDeviceAccessInformation's
+ * createFromDeviceClassId: (the passed in implementation should be the
+ * value returend from setupMock).
+ */
++ (void)tearDownMock:(IMP)implementation;
+
+/**
+ * Returns a mocked version of WDEDeviceAccessInformation.
+ */
++ (WDEDeviceAccessInformation*)mockCreateFromDeviceClassId:(WFGUID*)deviceClassId;
+
+@end
+
+@implementation __ABAddressBook_Swizzle
+
++ (IMP)setupMock {
+    SEL originalSelector = @selector(createFromDeviceClassId:);
+    SEL swizzledSelector = @selector(mockCreateFromDeviceClassId:);
+    Method swizzledMethod = class_getClassMethod([__ABAddressBook_Swizzle class], swizzledSelector);
+    return class_replaceMethod(objc_getMetaClass("WDEDeviceAccessInformation"),
+                               originalSelector,
+                               method_getImplementation(swizzledMethod),
+                               @encode(WFGUID*));
+}
+
++ (void)tearDownMock:(IMP)implementation {
+    class_replaceMethod(objc_getMetaClass("WDEDeviceAccessInformation"),
+                        @selector(createFromDeviceClassId:),
+                        implementation,
+                        @encode(WFGUID*));
+}
+
++ (WDEDeviceAccessInformation*)mockCreateFromDeviceClassId:(WFGUID*)deviceClassId {
+    auto fakeDeviceAccessInformation = Make<MockDeviceAccessInformation>();
+
+    fakeDeviceAccessInformation->Setget_CurrentStatus([](DeviceAccessStatus* result) {
+        *result = static_cast<DeviceAccessStatus>(statusToMock);
+        return S_OK;
+    });
+
+    return [WDEDeviceAccessInformation createWith:fakeDeviceAccessInformation.Get()];
+}
+
+@end
+
+TEST(AddressBook, AddressBook_AuthorizationStatus) {
+    IMP originalImplementation = [__ABAddressBook_Swizzle setupMock];
+
+    statusToMock = WDEDeviceAccessStatusUnspecified;
+    ASSERT_EQ_MSG(kABAuthorizationStatusNotDetermined,
+                  ABAddressBookGetAuthorizationStatus(),
+                  "FAILED: Authorization Status should be undetermined!\n");
+
+    statusToMock = WDEDeviceAccessStatusAllowed;
+    ASSERT_EQ_MSG(kABAuthorizationStatusAuthorized,
+                  ABAddressBookGetAuthorizationStatus(),
+                  "FAILED: Authorization Status should be authorized!\n");
+
+    statusToMock = WDEDeviceAccessStatusDeniedByUser;
+    ASSERT_EQ_MSG(kABAuthorizationStatusDenied, ABAddressBookGetAuthorizationStatus(), "FAILED: Authorization Status should be denied!\n");
+
+    statusToMock = WDEDeviceAccessStatusDeniedBySystem;
+    ASSERT_EQ_MSG(kABAuthorizationStatusRestricted,
+                  ABAddressBookGetAuthorizationStatus(),
+                  "FAILED: Authorization Status should be restricted!\n");
+
+    [__ABAddressBook_Swizzle tearDownMock:originalImplementation];
```

Comment:
>[__ABAddressBook_Swizzle tearDownMock:originalImplementation]; [](start = 4, length = 62)

setup and teardown methods are hard to get correct. For example what happens if your test encounters an exception? what about if it AVs (seg faults)? What state are we corrupting and how it will impact future tests? In general we want to isolate the test code itself from any setup / cleanup that needs to occur. In other test frameworks like TAEF, there is a facility known as test fixtures (for example TEST_CLASS_SETUP) that handle this nicely. You should investigate if something similar exists in gtest that you can use for this purpose.